File: src/lib/tauri/commands.ts
import { invoke } from "@tauri-apps/api/core";
import type { AudioLevelEvent } from "$lib/domain/audio-meter";
import type { CaptionSettings } from "$lib/domain/caption-settings";
import type {
  AsrDiagnosticResult,
  ModelSettingsStore,
  ModelStatus,
} from "$lib/domain/model-settings";
import type {
  OverlayProfileId,
  OverlayPlacement,
  OverlaySettings,
  OverlaySettingsStore,
} from "$lib/domain/overlay-settings";
import type { PerformanceSettings } from "$lib/domain/performance-settings";
import type {
  AudioSource,
  PlatformCapabilities,
  SourcePreview,
} from "$lib/domain/source-selection";
import type {
  TranscriptExportFormat,
  TranscriptSegment,
  TranscriptSettings,
  TranscriptSessionSummary,
} from "$lib/domain/transcript-settings";
import type {
  TranslationDiagnosticResult,
  TranslationEngineStatus,
  TranslationSettings,
} from "$lib/domain/translation-settings";

export type SessionState = "idle" | "starting" | "listening" | "error";

export type AppStatus = {
  state: SessionState;
  message: string;
  selectedSource: string | null;
};

export type AppMetadata = {
  productName: string;
  version: string;
};

export function getAppMetadata(): Promise<AppMetadata> {
  return invoke<AppMetadata>("get_app_metadata");
}

export function getAppStatus(): Promise<AppStatus> {
  return invoke<AppStatus>("get_app_status");
}

export function startAudioMeter(
  sourceIds: string[],
  sourceLabel?: string,
): Promise<void> {
  return invoke<void>("start_audio_meter", { sourceIds, sourceLabel });
}

export function stopAudioMeter(): Promise<void> {
  return invoke<void>("stop_audio_meter");
}

export function openCaptionWindow(): Promise<void> {
  return invoke<void>("open_caption_window");
}

export function closeCaptionWindow(): Promise<void> {
  return invoke<void>("close_caption_window");
}

export function isCaptionWindowOpen(): Promise<boolean> {
  return invoke<boolean>("is_caption_window_open");
}

export function getOverlaySettings(): Promise<OverlaySettings> {
  return invoke<OverlaySettings>("get_overlay_settings");
}

export function saveOverlaySettings(
  settings: OverlaySettings,
): Promise<OverlaySettings> {
  return invoke<OverlaySettings>("save_overlay_settings", { settings });
}

export function getOverlaySettingsStore(): Promise<OverlaySettingsStore> {
  return invoke<OverlaySettingsStore>("get_overlay_settings_store");
}

export function saveOverlaySettingsStore(
  store: OverlaySettingsStore,
): Promise<OverlaySettingsStore> {
  return invoke<OverlaySettingsStore>("save_overlay_settings_store", { store });
}

export function selectOverlaySettingsProfile(
  profileId: OverlayProfileId,
): Promise<OverlaySettingsStore> {
  return invoke<OverlaySettingsStore>("select_overlay_settings_profile", {
    profileId,
  });
}

export function startOverlayPlacement(
  settings: OverlaySettings,
): Promise<OverlayPlacement> {
  return invoke<OverlayPlacement>("start_overlay_placement", { settings });
}

export function getOverlayPlacement(): Promise<OverlayPlacement | null> {
  return invoke<OverlayPlacement | null>("get_overlay_placement");
}

export function stopOverlayPlacement(): Promise<void> {
  return invoke<void>("stop_overlay_placement");
}

export type { AudioLevelEvent };

export function getPlatformCapabilities(): Promise<PlatformCapabilities> {
  return invoke<PlatformCapabilities>("get_platform_capabilities");
}

export function listAvailableSources(): Promise<AudioSource[]> {
  return invoke<AudioSource[]>("list_available_sources");
}

export function getSourcePreviews(): Promise<SourcePreview[]> {
  return invoke<SourcePreview[]>("get_source_previews");
}

export function getModelSettingsStore(): Promise<ModelSettingsStore> {
  return invoke<ModelSettingsStore>("get_model_settings_store");
}

export function saveModelSettingsStore(
  store: ModelSettingsStore,
): Promise<ModelSettingsStore> {
  return invoke<ModelSettingsStore>("save_model_settings_store", { store });
}

export function getModelStatus(): Promise<ModelStatus> {
  return invoke<ModelStatus>("get_model_status");
}

export function installDefaultAsrAssets(): Promise<ModelSettingsStore> {
  return invoke<ModelSettingsStore>("install_default_asr_assets");
}

export function runAsrModelDiagnostic(): Promise<AsrDiagnosticResult> {
  return invoke<AsrDiagnosticResult>("run_asr_model_diagnostic");
}

export function runCaptionFlowDiagnostic(): Promise<AsrDiagnosticResult> {
  return invoke<AsrDiagnosticResult>("run_caption_flow_diagnostic");
}

export function getCaptionSettings(): Promise<CaptionSettings> {
  return invoke<CaptionSettings>("get_caption_settings");
}

export function saveCaptionSettings(
  settings: CaptionSettings,
): Promise<CaptionSettings> {
  return invoke<CaptionSettings>("save_caption_settings", { settings });
}

export function getTranslationSettings(): Promise<TranslationSettings> {
  return invoke<TranslationSettings>("get_translation_settings");
}

export function saveTranslationSettings(
  settings: TranslationSettings,
): Promise<TranslationSettings> {
  return invoke<TranslationSettings>("save_translation_settings", { settings });
}

export function getTranslationEngineStatus(): Promise<TranslationEngineStatus> {
  return invoke<TranslationEngineStatus>("get_translation_engine_status");
}

export function runTranslationEngineDiagnostic(): Promise<TranslationDiagnosticResult> {
  return invoke<TranslationDiagnosticResult>(
    "run_translation_engine_diagnostic",
  );
}

export function getPerformanceSettings(): Promise<PerformanceSettings> {
  return invoke<PerformanceSettings>("get_performance_settings");
}

export function savePerformanceSettings(
  settings: PerformanceSettings,
): Promise<PerformanceSettings> {
  return invoke<PerformanceSettings>("save_performance_settings", { settings });
}

export function getTranscriptSettings(): Promise<TranscriptSettings> {
  return invoke<TranscriptSettings>("get_transcript_settings");
}

export function saveTranscriptSettings(
  settings: TranscriptSettings,
): Promise<TranscriptSettings> {
  return invoke<TranscriptSettings>("save_transcript_settings", { settings });
}

export function startTranscriptSession(
  sourceSummary: string,
): Promise<number | null> {
  return invoke<number | null>("start_transcript_session", { sourceSummary });
}

export function appendTranscriptSegment(
  segment: TranscriptSegment,
): Promise<void> {
  return invoke<void>("append_transcript_segment", { segment });
}

export function finishTranscriptSession(): Promise<void> {
  return invoke<void>("finish_transcript_session");
}

export function listTranscriptSessions(
  limit = 20,
): Promise<TranscriptSessionSummary[]> {
  return invoke<TranscriptSessionSummary[]>("list_transcript_sessions", {
    limit,
  });
}

export function exportTranscriptSession(
  sessionId: number,
  format: TranscriptExportFormat,
): Promise<string> {
  return invoke<string>("export_transcript_session", { sessionId, format });
}

export function commandErrorMessage(error: unknown): string {
  if (error instanceof Error) {
    return error.message;
  }

  if (typeof error === "object" && error !== null) {
    const message = Reflect.get(error, "message");
    if (typeof message === "string") {
      return message;
    }
  }

  if (typeof error === "string" && error.length > 0) {
    return error;
  }

  return "The app could not complete that action.";
}

File: src-tauri/src/commands.rs
use crate::{
    app_error::{AppError, CommandResult},
    app_state::AppState,
    asr::{run_asr_diagnostic, run_system_audio_asr_diagnostic, AsrDiagnosticResult},
    caption_settings::CaptionSettings,
    model_settings::{ModelSettingsStore, ModelStatus},
    overlay_settings::{OverlayPlacement, OverlayProfileId, OverlaySettings, OverlaySettingsStore},
    performance_settings::PerformanceSettings,
    platform,
    platform::{PlatformCapabilityProvider, SourceEnumerator, SourcePreviewProvider},
    source::{AudioSource, PlatformCapabilities, SourcePreview},
    transcript::{
        TranscriptExportFormat, TranscriptSegment, TranscriptSessionSummary, TranscriptSettings,
    },
    translation::{
        run_translation_diagnostic, TranslationDiagnosticResult, TranslationEngineStatus,
        TranslationSettings,
    },
};
use serde::Serialize;
use tauri::{AppHandle, State};

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AppStatus {
    pub state: SessionState,
    pub message: String,
    pub selected_source: Option<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AppMetadata {
    pub product_name: String,
    pub version: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum SessionState {
    Idle,
    Starting,
    Listening,
    Error,
}

#[tauri::command]
pub fn get_app_metadata() -> AppMetadata {
    AppMetadata {
        product_name: "FeelSay".to_string(),
        version: env!("CARGO_PKG_VERSION").to_string(),
    }
}

#[tauri::command]
pub fn get_app_status() -> AppStatus {
    AppStatus {
        state: SessionState::Idle,
        message: "Ready to configure captions.".to_string(),
        selected_source: None,
    }
}

#[tauri::command]
pub fn start_audio_meter(
    app: AppHandle,
    state: State<'_, AppState>,
    source_ids: Vec<String>,
    source_label: Option<String>,
) -> CommandResult<()> {
    let asr_config = state.asr_runtime_config_for_source(source_label)?;

    state
        .audio_meter()
        .start_mock_meter(app, source_ids, asr_config)?;
    Ok(())
}

#[tauri::command]
pub fn stop_audio_meter(state: State<'_, AppState>) -> CommandResult<()> {
    state.audio_meter().stop()?;
    Ok(())
}

#[tauri::command]
pub fn open_caption_window(state: State<'_, AppState>) -> CommandResult<()> {
    state.open_caption_process()?;
    Ok(())
}

#[tauri::command]
pub fn close_caption_window(state: State<'_, AppState>) -> CommandResult<()> {
    state
        .close_caption_process()
        .map_err(|error| AppError::Window(error.to_string()))?;
    Ok(())
}

#[tauri::command]
pub fn is_caption_window_open(state: State<'_, AppState>) -> CommandResult<bool> {
    Ok(state
        .caption_process_is_running()
        .map_err(|error| AppError::Window(error.to_string()))?)
}

#[tauri::command]
pub fn get_overlay_settings(state: State<'_, AppState>) -> CommandResult<OverlaySettings> {
    Ok(state.overlay_settings().load()?)
}

#[tauri::command]
pub fn save_overlay_settings(
    state: State<'_, AppState>,
    settings: OverlaySettings,
) -> CommandResult<OverlaySettings> {
    Ok(state.overlay_settings().save(settings)?)
}

#[tauri::command]
pub fn get_overlay_settings_store(
    state: State<'_, AppState>,
) -> CommandResult<OverlaySettingsStore> {
    Ok(state.overlay_settings().load_store()?)
}

#[tauri::command]
pub fn save_overlay_settings_store(
    state: State<'_, AppState>,
    store: OverlaySettingsStore,
) -> CommandResult<OverlaySettingsStore> {
    Ok(state.overlay_settings().save_store(store)?)
}

#[tauri::command]
pub fn select_overlay_settings_profile(
    state: State<'_, AppState>,
    profile_id: OverlayProfileId,
) -> CommandResult<OverlaySettingsStore> {
    Ok(state.overlay_settings().select_profile(profile_id)?)
}

#[tauri::command]
pub fn start_overlay_placement(
    state: State<'_, AppState>,
    settings: OverlaySettings,
) -> CommandResult<OverlayPlacement> {
    Ok(state.start_overlay_placement(settings)?)
}

#[tauri::command]
pub fn get_overlay_placement(
    state: State<'_, AppState>,
) -> CommandResult<Option<OverlayPlacement>> {
    Ok(state.get_overlay_placement()?)
}

#[tauri::command]
pub fn stop_overlay_placement(state: State<'_, AppState>) -> CommandResult<()> {
    state
        .stop_overlay_placement()
        .map_err(|error| AppError::Window(error.to_string()))?;
    Ok(())
}

#[tauri::command]
pub fn get_platform_capabilities() -> PlatformCapabilities {
    platform::current_provider().capabilities()
}

#[tauri::command]
pub fn list_available_sources() -> CommandResult<Vec<AudioSource>> {
    Ok(platform::current_provider().list_sources()?)
}

#[tauri::command]
pub fn get_source_previews() -> CommandResult<Vec<SourcePreview>> {
    Ok(platform::current_provider().source_previews()?)
}

#[tauri::command]
pub fn get_model_settings_store(state: State<'_, AppState>) -> CommandResult<ModelSettingsStore> {
    Ok(state.model_settings().load()?)
}

#[tauri::command]
pub fn save_model_settings_store(
    state: State<'_, AppState>,
    store: ModelSettingsStore,
) -> CommandResult<ModelSettingsStore> {
    Ok(state.model_settings().save(store)?)
}

#[tauri::command]
pub fn get_model_status(state: State<'_, AppState>) -> CommandResult<ModelStatus> {
    Ok(state.model_settings().status()?)
}

#[tauri::command]
pub fn install_default_asr_assets(state: State<'_, AppState>) -> CommandResult<ModelSettingsStore> {
    Ok(state.model_settings().install_default_assets()?)
}

#[tauri::command]
pub fn run_asr_model_diagnostic(state: State<'_, AppState>) -> CommandResult<AsrDiagnosticResult> {
    let Some(config) = state.asr_runtime_config()? else {
        return Err(AppError::Asr("ASR model is not ready".to_string()).into());
    };

    Ok(run_asr_diagnostic(config)?)
}

#[tauri::command]
pub fn run_caption_flow_diagnostic(
    state: State<'_, AppState>,
) -> CommandResult<AsrDiagnosticResult> {
    let Some(config) = state.asr_runtime_config()? else {
        return Err(AppError::Asr("ASR model is not ready".to_string()).into());
    };

    state.open_caption_process()?;
    let result = run_system_audio_asr_diagnostic(config);
    state
        .close_caption_process()
        .map_err(|error| AppError::Window(error.to_string()))?;

    Ok(result?)
}

#[tauri::command]
pub fn get_caption_settings(state: State<'_, AppState>) -> CommandResult<CaptionSettings> {
    Ok(state.caption_settings().load()?)
}

#[tauri::command]
pub fn save_caption_settings(
    state: State<'_, AppState>,
    settings: CaptionSettings,
) -> CommandResult<CaptionSettings> {
    Ok(state.caption_settings().save(settings)?)
}

#[tauri::command]
pub fn get_translation_settings(state: State<'_, AppState>) -> CommandResult<TranslationSettings> {
    Ok(state.translation_settings().load()?)
}

#[tauri::command]
pub fn save_translation_settings(
    state: State<'_, AppState>,
    settings: TranslationSettings,
) -> CommandResult<TranslationSettings> {
    Ok(state.translation_settings().save(settings)?)
}

#[tauri::command]
pub fn get_translation_engine_status(state: State<'_, AppState>) -> TranslationEngineStatus {
    state.translation_settings().status()
}

#[tauri::command]
pub fn run_translation_engine_diagnostic(
    state: State<'_, AppState>,
) -> CommandResult<TranslationDiagnosticResult> {
    let caption_settings = state.caption_settings().load()?;
    let translation_settings = state.translation_settings().load()?;

    Ok(run_translation_diagnostic(
        translation_settings,
        caption_settings.translation_source_language,
        caption_settings.translation_target_language,
    )?)
}

#[tauri::command]
pub fn get_performance_settings(state: State<'_, AppState>) -> CommandResult<PerformanceSettings> {
    Ok(state.performance_settings().load()?)
}

#[tauri::command]
pub fn save_performance_settings(
    state: State<'_, AppState>,
    settings: PerformanceSettings,
) -> CommandResult<PerformanceSettings> {
    Ok(state.performance_settings().save(settings)?)
}

#[tauri::command]
pub fn get_transcript_settings(state: State<'_, AppState>) -> CommandResult<TranscriptSettings> {
    Ok(state.transcripts().load_settings()?)
}

#[tauri::command]
pub fn save_transcript_settings(
    state: State<'_, AppState>,
    settings: TranscriptSettings,
) -> CommandResult<TranscriptSettings> {
    Ok(state.transcripts().save_settings(settings)?)
}

#[tauri::command]
pub fn start_transcript_session(
    state: State<'_, AppState>,
    source_summary: String,
) -> CommandResult<Option<i64>> {
    Ok(state.transcripts().start_session(source_summary)?)
}

#[tauri::command]
pub fn append_transcript_segment(
    state: State<'_, AppState>,
    segment: TranscriptSegment,
) -> CommandResult<()> {
    state.transcripts().append_segment(segment)?;
    Ok(())
}

#[tauri::command]
pub fn finish_transcript_session(state: State<'_, AppState>) -> CommandResult<()> {
    state.transcripts().finish_session()?;
    Ok(())
}

#[tauri::command]
pub fn list_transcript_sessions(
    state: State<'_, AppState>,
    limit: u32,
) -> CommandResult<Vec<TranscriptSessionSummary>> {
    Ok(state.transcripts().list_sessions(limit)?)
}

#[tauri::command]
pub fn export_transcript_session(
    state: State<'_, AppState>,
    session_id: i64,
    format: TranscriptExportFormat,
) -> CommandResult<String> {
    Ok(state.transcripts().export_session(session_id, format)?)
}
